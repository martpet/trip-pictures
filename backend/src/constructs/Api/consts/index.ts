export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
