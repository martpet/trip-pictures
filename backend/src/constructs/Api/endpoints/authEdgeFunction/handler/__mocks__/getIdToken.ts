export const getIdToken = vi
  .fn()
  .mockName('getIdToken')
  .mockResolvedValue('dummyIdToken');
