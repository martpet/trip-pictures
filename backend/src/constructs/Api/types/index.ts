export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
