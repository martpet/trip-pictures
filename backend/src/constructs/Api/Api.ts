import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
