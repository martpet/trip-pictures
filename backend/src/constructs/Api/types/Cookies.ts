import { ConditionalKeys } from 'type-fest';

import { apiOptions } from '~/consts';
import { ApiRouteOptions } from '~/types';

export type PathWithCookies = ConditionalKeys<
  typeof apiOptions,
  Pick<ApiRouteOptions, 'cookies'>
>;

export type Cookies<T extends PathWithCookies> = Partial<
  Record<(typeof apiOptions)[T]['cookies'][number], string>
>;

export type AllCookies = Cookies<PathWithCookies>;
