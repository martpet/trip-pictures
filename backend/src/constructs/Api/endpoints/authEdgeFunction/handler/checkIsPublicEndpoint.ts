import { isPublicEndpoint } from '~/../../shared/utils/isPublicEndpoint';
import { LambdaEdgeViewerEvent } from '~/constructs/Api/types';

export function checkIsPublicEndpoint(edgeEvent: LambdaEdgeViewerEvent) {
  const { uri, method } = edgeEvent.Records[0].cf.request;

  return isPublicEndpoint({
    path: uri,
    method,
  });
}
