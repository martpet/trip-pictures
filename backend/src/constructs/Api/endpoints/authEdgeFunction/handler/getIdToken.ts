import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';

import { getIdTokenPayload } from '~/constructs/Api/utils/getIdTokenPayload';
import { refreshTokenExpiredErrorMessage, region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

import { fetchNewIdToken } from './fetchNewIdToken';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const getIdToken = async (sessionId: string) => {
  const getCommand = new GetCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
  });

  const { Item } = await ddbDocClient.send(getCommand);
  const sessionsItem = Item as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  const { idToken, refreshToken, refreshTokenExpires } = sessionsItem;
  const { exp: idTokenExpires, aud: clientId } = await getIdTokenPayload(idToken);

  if ((idTokenExpires - 5) * 1000 > Date.now()) {
    return idToken;
  }

  if (refreshTokenExpires * 1000 > Date.now()) {
    return fetchNewIdToken({ refreshToken, sessionId, clientId });
  }

  throw new Error(refreshTokenExpiredErrorMessage);
};
