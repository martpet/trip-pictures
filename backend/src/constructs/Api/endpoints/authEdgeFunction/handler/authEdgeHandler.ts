import clone from 'clone';
import { StatusCodes } from 'http-status-codes';

import { LambdaEdgeViewerRequestHandler } from '~/constructs/Api/types';
import { edgeErrorResponse } from '~/constructs/Api/utils/edgeErrorResponse';
import { parseLambdaEdgeEventCookies } from '~/constructs/Api/utils/parseLambdaEdgeEventCookies';

import { checkIsPublicEndpoint } from './checkIsPublicEndpoint';
import { getIdToken } from './getIdToken';

export const handler: LambdaEdgeViewerRequestHandler = async (event) => {
  try {
    const { request } = event.Records[0].cf;
    delete request.headers.authorization;

    if (request.method === 'OPTIONS') {
      return { status: StatusCodes.NO_CONTENT };
    }

    if (checkIsPublicEndpoint(event)) {
      return request;
    }

    const { sessionId } = parseLambdaEdgeEventCookies(event);

    if (!sessionId) {
      throw new Error('invalid or missing `sessionId` cookie');
    }

    const requestClone = clone(request);
    requestClone.headers.authorization = [{ value: await getIdToken(sessionId) }];
    return requestClone;
  } catch (error) {
    return edgeErrorResponse('eWUjcKPBi3', {
      statusCode: StatusCodes.UNAUTHORIZED,
      error,
    });
  }
};
