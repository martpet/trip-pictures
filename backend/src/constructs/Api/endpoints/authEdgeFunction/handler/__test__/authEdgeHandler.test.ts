import { lambdaEdgeViewerEvent } from '~/constructs/Api/consts';
import { LambdaEdgeViewerRequestHandler } from '~/constructs/Api/types';
import {
  itCalls,
  itResolves,
  itResolvesWithEdgeError,
  parseLambdaEdgeEventCookies,
} from '~/constructs/Api/utils';

import { handler } from '../authEdgeHandler';
import { checkIsPublicEndpoint } from '../checkIsPublicEndpoint';
import { getIdToken } from '../getIdToken';

vi.mock('../checkIsPublicEndpoint');
vi.mock('../getIdToken');
vi.mock('~/constructs/Api/utils/parseLambdaEdgeEventCookies');
vi.mock('~/constructs/Api/utils/edgeErrorResponse');

const args = [lambdaEdgeViewerEvent] as Parameters<LambdaEdgeViewerRequestHandler>;

vi.mocked(checkIsPublicEndpoint).mockReturnValue(false);
vi.mocked(parseLambdaEdgeEventCookies).mockReturnValue({
  sessionId: 'dummySessionId',
});

describe('authEdgeHandler', () => {
  itResolves(handler, args);
  itCalls(parseLambdaEdgeEventCookies, handler, args);
  itCalls(checkIsPublicEndpoint, handler, args);
  itCalls(getIdToken, handler, args);

  describe('when the endpoint is public', () => {
    beforeEach(() => {
      vi.mocked(checkIsPublicEndpoint).mockReturnValueOnce(true);
    });
    itResolves(handler, args);
  });

  describe('when the request method is "OPTIONS"', () => {
    const argsClone = structuredClone(args);
    argsClone[0].Records[0].cf.request.method = 'OPTIONS';
    itResolves(handler, argsClone);
  });

  describe('when "sessionId" is missing', () => {
    beforeEach(() => {
      vi.mocked(parseLambdaEdgeEventCookies).mockReturnValueOnce({});
    });
    itResolvesWithEdgeError(handler, args);
  });

  describe('when "getIdToken" rejects', () => {
    beforeEach(() => {
      vi.mocked(getIdToken).mockRejectedValueOnce(
        new Error('dummy getIdToken error message')
      );
    });
    itResolvesWithEdgeError(handler, args);
  });
});
