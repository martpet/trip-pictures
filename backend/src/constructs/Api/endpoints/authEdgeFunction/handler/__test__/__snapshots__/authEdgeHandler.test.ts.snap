// Vitest Snapshot v1

exports[`authEdgeHandler > calls "checkIsPublicEndpoint" with correct args 1`] = `
[
  [
    {
      "Records": [
        {
          "cf": {
            "config": {
              "distributionDomainName": "d111111abcdef8.cloudfront.net",
              "distributionId": "EDFDVBD6EXAMPLE",
              "eventType": "viewer-request",
              "requestId": "4TyzHTaYWb1GX1qTfsHhEqV6HUDd_BzoBZnwfnvQc_1oF26ClkoUSEQ==",
            },
            "request": {
              "clientIp": "203.0.113.178",
              "headers": {
                "accept": [
                  {
                    "key": "accept",
                    "value": "*/*",
                  },
                ],
                "cookie": [
                  {
                    "key": "cookie",
                    "value": "dummyCookieValue",
                  },
                ],
                "host": [
                  {
                    "key": "Host",
                    "value": "d111111abcdef8.cloudfront.net",
                  },
                ],
                "user-agent": [
                  {
                    "key": "User-Agent",
                    "value": "curl/7.66.0",
                  },
                ],
              },
              "method": "GET",
              "querystring": "",
              "uri": "/",
            },
          },
        },
      ],
    },
  ],
]
`;

exports[`authEdgeHandler > calls "getIdToken" with correct args 1`] = `
[
  [
    "dummySessionId",
  ],
]
`;

exports[`authEdgeHandler > calls "parseLambdaEdgeEventCookies" with correct args 1`] = `
[
  [
    {
      "Records": [
        {
          "cf": {
            "config": {
              "distributionDomainName": "d111111abcdef8.cloudfront.net",
              "distributionId": "EDFDVBD6EXAMPLE",
              "eventType": "viewer-request",
              "requestId": "4TyzHTaYWb1GX1qTfsHhEqV6HUDd_BzoBZnwfnvQc_1oF26ClkoUSEQ==",
            },
            "request": {
              "clientIp": "203.0.113.178",
              "headers": {
                "accept": [
                  {
                    "key": "accept",
                    "value": "*/*",
                  },
                ],
                "cookie": [
                  {
                    "key": "cookie",
                    "value": "dummyCookieValue",
                  },
                ],
                "host": [
                  {
                    "key": "Host",
                    "value": "d111111abcdef8.cloudfront.net",
                  },
                ],
                "user-agent": [
                  {
                    "key": "User-Agent",
                    "value": "curl/7.66.0",
                  },
                ],
              },
              "method": "GET",
              "querystring": "",
              "uri": "/",
            },
          },
        },
      ],
    },
  ],
]
`;

exports[`authEdgeHandler > resolves with a correct value 1`] = `
{
  "clientIp": "203.0.113.178",
  "headers": {
    "accept": [
      {
        "key": "accept",
        "value": "*/*",
      },
    ],
    "authorization": [
      {
        "value": "dummyIdToken",
      },
    ],
    "cookie": [
      {
        "key": "cookie",
        "value": "dummyCookieValue",
      },
    ],
    "host": [
      {
        "key": "Host",
        "value": "d111111abcdef8.cloudfront.net",
      },
    ],
    "user-agent": [
      {
        "key": "User-Agent",
        "value": "curl/7.66.0",
      },
    ],
  },
  "method": "GET",
  "querystring": "",
  "uri": "/",
}
`;

exports[`authEdgeHandler > when "getIdToken" rejects > calls "edgeErrorResponse" with correct args 1`] = `
[
  [
    "eWUjcKPBi3",
    {
      "error": [Error: dummy getIdToken error message],
      "statusCode": 401,
    },
  ],
]
`;

exports[`authEdgeHandler > when "getIdToken" rejects > resolves with a correct value 1`] = `undefined`;

exports[`authEdgeHandler > when "sessionId" is missing > calls "edgeErrorResponse" with correct args 1`] = `
[
  [
    "eWUjcKPBi3",
    {
      "error": [Error: invalid or missing \`sessionId\` cookie],
      "statusCode": 401,
    },
  ],
]
`;

exports[`authEdgeHandler > when "sessionId" is missing > resolves with a correct value 1`] = `undefined`;

exports[`authEdgeHandler > when the endpoint is public > resolves with a correct value 1`] = `
{
  "clientIp": "203.0.113.178",
  "headers": {
    "accept": [
      {
        "key": "accept",
        "value": "*/*",
      },
    ],
    "cookie": [
      {
        "key": "cookie",
        "value": "dummyCookieValue",
      },
    ],
    "host": [
      {
        "key": "Host",
        "value": "d111111abcdef8.cloudfront.net",
      },
    ],
    "user-agent": [
      {
        "key": "User-Agent",
        "value": "curl/7.66.0",
      },
    ],
  },
  "method": "GET",
  "querystring": "",
  "uri": "/",
}
`;

exports[`authEdgeHandler > when the request method is "OPTIONS" > resolves with a correct value 1`] = `
{
  "status": 204,
}
`;
